src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3657,12 +3657,36 @@ void MacroAssembler::compiler_fast_unlock_object(Register oop, Register box, Reg
 
   bind(not_recursive);
 
+  NearLabel check_succ, set_eq_unlocked;
+
+  // Set owner to null.
+  z_release();
+  z_lghi(temp, 0);
+  z_stg(temp, OM_OFFSET_NO_MONITOR_VALUE_TAG(owner), currentHeader);
+  z_fence(); // membar(StoreLoad);
+
+  // Check if the entry lists are empty.
   load_and_test_long(temp, Address(currentHeader, OM_OFFSET_NO_MONITOR_VALUE_TAG(EntryList)));
-  z_brne(done);
+  z_brne(check_succ);
   load_and_test_long(temp, Address(currentHeader, OM_OFFSET_NO_MONITOR_VALUE_TAG(cxq)));
-  z_brne(done);
-  z_release();
-  z_stg(temp/*=0*/, OM_OFFSET_NO_MONITOR_VALUE_TAG(owner), currentHeader);
+  z_bre(done); // If so we are done.
+
+  bind(check_succ);
+
+  // Check if there is a successor.
+  load_and_test_long(temp, Address(currentHeader, OM_OFFSET_NO_MONITOR_VALUE_TAG(succ)));
+  z_brne(set_eq_unlocked); // If so we are done.
+
+  // Save the monitor pointer in the current thread, so we can try to
+  // reacquire the lock in SharedRuntime::monitor_exit_helper().
+  z_xilf(currentHeader, markWord::monitor_value);
+  z_stg(currentHeader, Address(Z_thread, JavaThread::unlocked_inflated_monitor_offset()));
+
+  z_cr(currentHeader, Z_thread); // Set flag = NE
```

Comment:
I ran tier1 test and don't see any new failure appearing. 

How about using `z_ltgr` here ? 

```suggestion
  z_ltgr(oop, oop); // Set flag = NE
```

Reply:
fixed